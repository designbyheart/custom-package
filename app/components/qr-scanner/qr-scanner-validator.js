import { schemaValidator } from '../../services/'
import {
  QR_CODE_SENDER_NAME,
  QR_CODE_UID,
  QR_CODE_REMOTE_HOSTING_DID,
  QR_CODE_REMOTE_PAIR_WISE_DID,
  QR_CODE_SIGNATURE,
  QR_CODE_TARGET_NAME,
  QR_CODE_CHALLENGE,
  QR_CODE_TARGET_DID,
} from '../../common/api-constants'

const qrSchema = {
  type: 'object',
  properties: {
    [QR_CODE_CHALLENGE]: { type: 'string' },
    [QR_CODE_SIGNATURE]: { type: 'string' },
  },
  required: [QR_CODE_CHALLENGE, QR_CODE_SIGNATURE],
}

const qrChallengeSchema = {
  type: 'object',
  properties: {
    [QR_CODE_UID]: { type: 'string' },
    [QR_CODE_REMOTE_HOSTING_DID]: { type: 'string' },
    [QR_CODE_REMOTE_PAIR_WISE_DID]: { type: 'string' },
    [QR_CODE_SENDER_NAME]: { type: 'string' },
    [QR_CODE_TARGET_NAME]: { type: 'string' },
    [QR_CODE_TARGET_DID]: { type: 'string' },
  },
  required: [
    QR_CODE_UID,
    QR_CODE_REMOTE_HOSTING_DID,
    QR_CODE_REMOTE_PAIR_WISE_DID,
    QR_CODE_SENDER_NAME,
    QR_CODE_TARGET_NAME,
    QR_CODE_TARGET_DID,
  ],
}

// this is the number that is max allowed by 200 X 200 qr code
// if we are getting more than that, then there is an issue from
// qr code generation site or someone is trying to break the app
const qrCodeAllowedLength = 1200

// If we need to use QR scanner somewhere else in app as well
// remove this method and create an equivalent in another
// component as well, for now this is good here
export function isValidQrCode(qr) {
  // check the length of qr data so that we don't get memory issues
  // while parsing the json
  if (qr.length >= qrCodeAllowedLength) {
    return false
  }

  let qrData
  try {
    qrData = JSON.parse(qr)
  } catch (e) {
    // qr code is not a valid json, we still return false
    return false
  }

  // check if qr code schema is valid
  if (!schemaValidator.validate(qrSchema, qrData)) {
    return false
  }

  let qrChallenge
  try {
    qrChallenge = JSON.parse(qrData.c)
  } catch (e) {
    return false
  }

  // check if qr code challenge schema is valid
  if (!schemaValidator.validate(qrChallengeSchema, qrChallenge)) {
    return false
  }

  // TODO:KS Check if signature is valid for qr code once libindy is integrated

  // since all checks are passed
  // qr code data is valid, send the parsed data back
  return {
    challenge: qrChallenge,
    signature: qrData.s,
    qrData,
  }
}
