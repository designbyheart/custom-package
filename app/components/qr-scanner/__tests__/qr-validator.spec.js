import { isValidQrCode } from '../qr-scanner-validator'

describe('QR Scanner validator', () => {
  it('should return false if schema is not valid at top level', () => {
    const invalidQrSchemas = [
      { a: '' },
      { c: '' },
      { c: '', a: '' },
      { s: '', a: '' },
      { s: '', 1: '' },
      // invalid as per length
      {
        c: '',
        s: 'a'.repeat(1300),
      },
      // invalid at second level
      { c: JSON.stringify({ n: '' }), s: '' },
      { c: JSON.stringify({ n: '', rci: '' }), s: '' },
      { c: JSON.stringify({ n: '', rci: '', en: '' }), s: '' },
      { c: JSON.stringify({ n: '', rci: '', en: '', un: '' }), s: '' },
      { c: JSON.stringify({ n: '', rci: '', en: '', un: '', iic: '' }), s: '' },
    ]

    invalidQrSchemas.map(schema =>
      expect(isValidQrCode(JSON.stringify(schema))).toBe(false)
    )
  })

  it('should return data for valid schema', () => {
    const validSchema = {
      c: JSON.stringify({
        tDID: '',
        sn: '',
        tn: '',
        uid: '',
        rhDID: '',
        rpDID: '',
      }),
      s: '',
    }

    expect(isValidQrCode(JSON.stringify(validSchema))).toMatchSnapshot()
  })
})
