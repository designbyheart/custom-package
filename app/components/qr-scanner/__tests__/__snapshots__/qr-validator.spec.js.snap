// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QR Scanner validator should return data for valid schema 1`] = `
Object {
  "challenge": Object {
    "rhDID": "",
    "rpDID": "",
    "sn": "",
    "tDID": "",
    "tn": "",
    "uid": "",
  },
  "qrData": Object {
    "c": "{\\"tDID\\":\\"\\",\\"sn\\":\\"\\",\\"tn\\":\\"\\",\\"uid\\":\\"\\",\\"rhDID\\":\\"\\",\\"rpDID\\":\\"\\"}",
    "s": "",
  },
  "signature": "",
}
`;
